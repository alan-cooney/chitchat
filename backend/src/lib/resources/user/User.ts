import { IGraphqlApi, MappingTemplate } from "@aws-cdk/aws-appsync";
import { AttributeType, BillingMode, Table } from "@aws-cdk/aws-dynamodb";
import { Construct } from "@aws-cdk/core";
import { join } from "path";

/**
 * User Functionality
 */
export default class User {
  constructor(scope: Construct, api: IGraphqlApi) {
    const id = "user";

    // Data
    const table = new Table(scope, `${id}Table`, {
      billingMode: BillingMode.PAY_PER_REQUEST,
      partitionKey: {
        name: "pk", // Will be APP#{{app-id}}#USER#{{user-id}}
        type: AttributeType.STRING,
      },
      pointInTimeRecovery: true,
    });

    const tableDS = api.addDynamoDbDataSource(`${id}TableDataSource`, table);

    // Mutation -> updateUser
    tableDS.createResolver({
      typeName: "Mutation",
      fieldName: "updateUser",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "./velocity/updateUserRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });

    // Query -> user
    tableDS.createResolver({
      typeName: "Query",
      fieldName: "user",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "./velocity/userRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });

    // Participant -> user
    tableDS.createResolver({
      typeName: "Participant",
      fieldName: "user",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "./velocity/participantUserRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });
  }
}
