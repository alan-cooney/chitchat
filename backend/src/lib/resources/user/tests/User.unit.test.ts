import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";
import { GraphqlApi } from "@aws-cdk/aws-appsync";
import User from "../User";

/**
 * Test stack
 */
const stack = new Stack();
const api = new GraphqlApi(stack, "api", {
  name: "test",
});
new User(stack, api);

it("creates a dynamodb table", () => {
  expect(stack).toHaveResource("AWS::DynamoDB::Table", {
    BillingMode: "PAY_PER_REQUEST",
    KeySchema: [
      {
        AttributeName: "pk",
        KeyType: "HASH",
      },
    ],
    PointInTimeRecoverySpecification: {
      PointInTimeRecoveryEnabled: true,
    },
  });
});

it("creates resources matching the snapshot", () => {
  const cloudformation = SynthUtils.toCloudFormation(stack);
  expect(cloudformation).toMatchSnapshot();
});
