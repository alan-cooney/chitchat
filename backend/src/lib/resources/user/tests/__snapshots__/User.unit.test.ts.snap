// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
Object {
  "Resources": Object {
    "apiC8550315": Object {
      "Properties": Object {
        "AuthenticationType": "API_KEY",
        "Name": "test",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "apiDefaultApiKey6AB8D7C4": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "apiSchema0EA92056": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "Definition": "",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "apiuserTableDataSourceA066116E": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DynamoDBConfig": Object {
          "AwsRegion": Object {
            "Ref": "AWS::Region",
          },
          "TableName": Object {
            "Ref": "userTableA39440AD",
          },
        },
        "Name": "userTableDataSource",
        "ServiceRoleArn": Object {
          "Fn::GetAtt": Array [
            "apiuserTableDataSourceServiceRole30F51588",
            "Arn",
          ],
        },
        "Type": "AMAZON_DYNAMODB",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "apiuserTableDataSourceMutationupdateUserResolver3BE1D95C": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apiuserTableDataSourceA066116E",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "userTableDataSource",
        "FieldName": "updateUser",
        "Kind": "UNIT",
        "RequestMappingTemplate": "#set($input = $context.args.input)

{
    \\"version\\": \\"2017-02-28\\",
    \\"operation\\": \\"PutItem\\",
    \\"key\\" : {
        \\"pk\\" : $util.dynamodb.toDynamoDBJson(\\"APP#$ctx.args.input.appID#USER#$ctx.args.input.id\\")
    },
    \\"attributeValues\\": $util.dynamodb.toMapValuesJson($input)
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apiuserTableDataSourceParticipantuserResolver7FFE0844": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apiuserTableDataSourceA066116E",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "userTableDataSource",
        "FieldName": "user",
        "Kind": "UNIT",
        "RequestMappingTemplate": "/home/alan/repos/chitchat/backend/src/lib/velocity/participantUserRequest.vtl",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Participant",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apiuserTableDataSourceQueryuserResolverCC9158BC": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apiuserTableDataSourceA066116E",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "userTableDataSource",
        "FieldName": "user",
        "Kind": "UNIT",
        "RequestMappingTemplate": "{
    \\"version\\": \\"2017-02-28\\",
    \\"operation\\": \\"GetItem\\",
    \\"key\\": {
        \\":pk\\": \\"APP#$ctx.args.appID#USER#$ctx.args.id\\"
    }
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apiuserTableDataSourceServiceRole30F51588": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "apiuserTableDataSourceServiceRoleDefaultPolicyB739E08E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "userTableA39440AD",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "apiuserTableDataSourceServiceRoleDefaultPolicyB739E08E",
        "Roles": Array [
          Object {
            "Ref": "apiuserTableDataSourceServiceRole30F51588",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "userTableA39440AD": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
